components/omega/doc/design/Driver.md

```diff
@@ -0,0 +1,253 @@
+```c++
+int OMEGA::OcnInit(
+   MPIComm Comm, ///< [in] ocean MPI communicator
+   OMEGA::TimeInstant &StartTime, ///< [out] sim start time
+   OMEGA::TimeInterval &RunInterval, ///< [out] interval for run method
+   OMEGA::State &CurrState, ///< [out] current model state
+   other args as needed
+);
+```
+The routine will return the initial state, the start time and
+the run interval computed based on input options. An integer
+return value will be non-zero if errors are encountered and zero
+if successful.
+It is also possible that a multi-stage initialization may
+be needed, especially in coupled mode and would require
+additional OcnInit interfaces. This will be determined during
+integration with E3SM later.
+
+#### 4.2.2 Run
+
+The run method will advance the model one time interval,
+typically the coupling time or the fastest forcing interval.
+The interface will look like:
+```c++
+int OMEGA::OcnRun(
+   OMEGA::TimeInstant &CurrTime, ///< [inout] current sim time
+   OMEGA::TimeInterval &RunInterval, ///< [in] interval to advance model
+   OMEGA::Alarm &EndAlarm, ///< [out] alarm to end simulation
+   OMEGA::State &CurrState, ///< [inout] current model state
+   other args as needed (eg forcing)
+)
+```
+The model state, current time and the time interval will be input.
+Other variables will be needed as well, like the surface forcing
+fields, and will be added as needed. On return, the time instant
+will contain the end time of the interval and the end alarm will
+be ringing if the end of the simulation has been reached. The
+CurrState will be the ocean state at the end of the run interval.
+An integer error code will be zero if successful and non-zero if an
+error was encountered. This interface will be modified as needed to
+include other fields.
+
+#### 4.2.3 Finalize
+
+The finalize method will write a checkpoint/restart file
+(if not already written by driver or run method on the
+final timestep) and then clean up all arrays and classes by
+calling the relevant routines for all Omega modules. The
+interface is similar to the Init interface:
+
+```c++
+int OMEGA::OcnFinalize(
+   OMEGA::TimeInstant &CurrTime, ///< [in] current sim time
+   OMEGA::State &CurrState, ///< [in] current model state
+   other args as needed for restart
+);
+```
+An integer return value will be zero if successful and
+non-zero if an error is encountered either writing a
+restart or deallocating memory.
+
+#### 4.2.4 Standalone driver (main)
+
+With the interfaces above, the standalone driver should look
+something like the code below (details subject to change during implementation).
+```c++
+
+int main(int argc, char **argv) {
+
+   MPI_Init(); // initialize MPI
+   yakl::init(); // initialize YAKL
+
+   OMEGA::State CurrState;
+   OMEGA::TimeInstant CurrTime;
+   OMEGA::TimeInterval RunInterval;
+   OMEGA::Alarm EndAlarm;
+
+   int Err = OcnInit(MPI_COMM_WORLD, CurrTime, RunInterval, CurrState,
+                     etc);
```

Comment:
Obviously the argument lists are meant to be precise, but I think `EndAlarm` could be added as an argument here (and correspondingly in 4.2.1) since it's defined above along with the other arguments that are explicitly listed.

Reply:
Added an EndAlarm argument to init for initializing the end alarm.